### PR TITLE
Add Istio strict/permissive auth e2e tests for Gloo Edge Gateway apis using new testing framework

### PR DESCRIPTION
**Pull request by npolshakova**
# Description

Should go in after https://github.com/solo-io/gloo/pull/9460 which enables CI to run Gloo Gateway API tests in ci. 

This PR adds:
- the Gloo Edge Gateway API setup for automtls enabled/disabled for strict/permissive peer auth
- Glooctl "istio inject"/"istio uninject" command added as a separate setup 

Follows up:
- Add "upgrade" path (Replicate old Istio regression test with first Upstream sslConfig configured, then remove Upstream sslConfig) with Gloo Edge Gateway API
- Add "downgrade" path (Automtls enabled, add sslConfig with old config, do a helm upgrade -i to update settings) with Gloo Edge Gateway API 
- Add Upstream sslConfig overwrite test for Gloo Edge Gateway API setup with disabled peer auth + automtls disabled on Upstream
- Add Upstream sslConfig overwrite test for k8s Gateway API setup with disabled peer auth + automtls disabled on Upstream

## Code changes

New Tests:
- Gloo Edge Gateway API suite with automtls enabled 
    - Headless svc test 
    - Strict/Permissive peerauth
- Gloo Edge Gateway API suite with automtls disabled 
    - Headless svc test 
    - Strict/Permissive peerauth
- Gloo Edge Gateway API suite with no Istio integration enabled. Istio integration is added via `glooctl inject/uninject` 
    - Strict/Permissive peerauth

New Assertions:
- `EventuallyPodsMatches` checks pod(s) match the input gomega matchers. A new matcher `PodHasContainersMatcher` is also added to check the gateway has the `sds` and `istio-proxy` after running `glooctl istio inject`. The inverse assertion is made after `glooctl istio uninject` is run. 

## CI Changes

Adds the following tests to cluster_two ci:

- TestGlooGatewayEdgeGateway
- TestGlooctlIstioInjectEdgeApiGateway
- TestAutomtlsIstioEdgeApisGateway
- TestIstioEdgeApiGateway

```
   go-test-run-regex: '(^TestK8sGatewayIstioAutoMtls$$|^TestK8sGatewayIstio$$|^TestGlooGatewayEdgeGateway$$|^TestGlooctlIstioInjectEdgeApiGateway$$|^TestAutomtlsIstioEdgeApisGateway$$|^TestIstioEdgeApiGateway$$)'
```

Run times are updated in follow up Upstream PR: https://github.com/solo-io/gloo/pull/9463 
- Updates run times for cluster one: https://github.com/solo-io/gloo/actions/runs/9065234123/job/24905385485?pr=9436 
- Update run time for cluster two: https://github.com/solo-io/gloo/actions/runs/9065234123/job/24905386213?pr=9436 

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: test/kubernetes/testutils/assertions/pods.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	"time"
+
+	"github.com/onsi/gomega"
+	"github.com/onsi/gomega/types"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EventuallyPodsMatches asserts that the pod(s) in the given namespace matches the provided matcher
+func (p *Provider) EventuallyPodsMatches(ctx context.Context, podNamespace string, listOpt metav1.ListOptions, matcher types.GomegaMatcher) {
```
Comment (npolshakova) [inline test/kubernetes/testutils/assertions/pods.go:15]: Pod assertion

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 2]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
 	}, extraArgs...)
 	return ExecuteCommandWithArgs(c.NewCommand(ctx), debugLogsArgs...)
 }
+
+// IstioInject inject istio-proxy and sds
+func (c *GlooCli) IstioInject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	checkArgs := append([]string{
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:70]:
```suggestion
	injectArgs := append([]string{
```

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 3]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
...
+		"--istio-namespace", "istio-system",
+		"--kube-context", kubeContext,
+	}, extraArgs...)
+
+	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), checkArgs...)
+}
+
+// IstioUninject uninjects istio-proxy and sds
+func (c *GlooCli) IstioUninject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	checkArgs := append([]string{
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:82]:
```suggestion
	uninjectArgs := append([]string{
```

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 4]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
...
+// IstioInject inject istio-proxy and sds
+func (c *GlooCli) IstioInject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	checkArgs := append([]string{
+		"istio", "inject",
+		"--namespace", installNamespace,
+		"--istio-namespace", "istio-system",
+		"--kube-context", kubeContext,
+	}, extraArgs...)
+
+	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), checkArgs...)
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:77]:
```suggestion
	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), injectArgs...)
```

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 5]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
...
+// IstioUninject uninjects istio-proxy and sds
+func (c *GlooCli) IstioUninject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	checkArgs := append([]string{
+		"istio", "inject",
+		"--namespace", installNamespace,
+		"--istio-namespace", "istio-system",
+		"--kube-context", kubeContext,
+	}, extraArgs...)
+
+	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), checkArgs...)
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:89]:
```suggestion
	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), uninjectArgs...)
```

File: test/kubernetes/e2e/features/glooctl/istio_inject_suite.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+func NewIstioInjectTestingSuite(ctx context.Context, testInst *e2e.TestInstallation) suite.TestingSuite {
+	return &istioInjectTestingSuite{
+		ctx:              ctx,
+		testInstallation: testInst,
+	}
+}
+
+func (s *istioInjectTestingSuite) TestCanInject() {
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/glooctl/istio_inject_suite.go:33]: This test will leave resources in a different state than it started (ie we've injected the sidecars). Is that ok, or should be clear it up in the TearDownSuite

File: test/kubernetes/e2e/features/glooctl/istio_uninject_suite.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+func NewIstioUninjectTestingSuite(ctx context.Context, testInst *e2e.TestInstallation) suite.TestingSuite {
+	return &istioUninjectTestingSuite{
+		ctx:              ctx,
+		testInstallation: testInst,
+	}
+}
+
+func (s *istioUninjectTestingSuite) TestCanUninject() {
+	// Uninject istio with glooctl
+	out, err := s.testInstallation.Actions.Glooctl().IstioUninject(s.ctx, s.testInstallation.Metadata.InstallNamespace, s.testInstallation.TestCluster.ClusterContext.KubeContext)
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/glooctl/istio_uninject_suite.go:35]: nit: It might be helpful to callout an assumption this test makes, which is that the inject test ran before it, right?

File: test/kubernetes/e2e/features/istio/gloo_gateway_no_mtls_suite.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
+package istio
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/istio/gloo_gateway_no_mtls_suite.go:1]: This file is named "no mtls suite", is that intentional?

File: test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		t.Fatalf("failed to install istio: %v", err)
+	}
+
+	// Install Gloo Gateway with only Edge APIs enabled
+	testInstallation.InstallGlooGateway(ctx, func(ctx context.Context) error {
+		return testHelper.InstallGloo(ctx, helper.GATEWAY, 5*time.Minute, helper.ExtraArgs("--values", testInstallation.Metadata.ValuesManifestFile))
+	})
+
+	// NOTE: Order of tests is important here because the tests are dependent on each other (e.g. the inject test must run before the istio test)
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go:69]: Woohoo!

File: test/kubernetes/e2e/gloo_gateway_edge/istio_edge_api_test.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+		t,
+		&gloogateway.Context{
+			InstallNamespace:   "istio-edge-api-gateway-test",
+			ValuesManifestFile: filepath.Join(util.MustGetThisDir(), "manifests", "istio-edge-api-gateway-test-helm.yaml"),
+		},
+	)
+
+	testHelper := e2e.MustTestHelper(ctx, testInstallation)
+
+	// create a tmp output directory for generated resources
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/gloo_gateway_edge/istio_edge_api_test.go:36]: We shouldn't need this anymore, right? Because the Testinstallation.GeneratedFiles.TempDir is what we can use

**Review by sam-heilbron — COMMENTED** [review]
Looking great! mostly small comments

File: test/kubernetes/e2e/features/glooctl/istio_inject_suite.go (hunk 6 above)
Comment (npolshakova) [reply test/kubernetes/e2e/features/glooctl/istio_inject_suite.go:33]: Yep, I'll add a comment to the top of the test but the inject/uninject needs to sandwich the Istio strict/permissive test in order for it to run.

File: test/kubernetes/e2e/features/istio/gloo_gateway_no_mtls_suite.go (hunk 8 above)
Comment (npolshakova) [reply test/kubernetes/e2e/features/istio/gloo_gateway_no_mtls_suite.go:1]: Yep, this is the Edge Gloo Gateway API version of the `k8s_gateway_no_mtls_suite.go` test. The manifest does not have auto mtls enabled. I can renamed the tests to be `k8s_gateway_no_auto_mtls_suite` and `gloo_gateway_no_auto_mtls_suite` to make it more clear!

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 11]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
 	}, extraArgs...)
 	return ExecuteCommandWithArgs(c.NewCommand(ctx), debugLogsArgs...)
 }
+
+// IstioInject inject istio-proxy and sds
+func (c *GlooCli) IstioInject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	injectArgs := append([]string{
+		"istio", "inject",
+		"--namespace", installNamespace,
+		"--istio-namespace", "istio-system",
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:73]: nit: We could use a constant here?

File: projects/gloo/cli/pkg/testutils/cli.go [hunk 12]
```diff
@@ -64,3 +64,27 @@ func (c *GlooCli) DebugLogs(ctx context.Context, extraArgs ...string) error {
...
+		"--kube-context", kubeContext,
+	}, extraArgs...)
+
+	return ExecuteCommandWithArgsOut(c.NewCommand(ctx), injectArgs...)
+}
+
+// IstioUninject uninjects istio-proxy and sds
+func (c *GlooCli) IstioUninject(ctx context.Context, installNamespace, kubeContext string, extraArgs ...string) (string, error) {
+	uninjectArgs := append([]string{
+		"istio", "inject",
```
Comment (sam-heilbron) [inline projects/gloo/cli/pkg/testutils/cli.go:83]:
```suggestion
		"istio", "uninject",
```

File: test/kubernetes/e2e/features/glooctl/istio_inject_suite.go [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+	// Inject istio with glooctl
+	out, err := s.testInstallation.Actions.Glooctl().IstioInject(s.ctx, s.testInstallation.Metadata.InstallNamespace, s.testInstallation.TestCluster.ClusterContext.KubeContext)
+	s.Assert().NoError(err, "Failed to inject istio")
+	s.Assert().Contains(out, "Istio injection was successful!")
+
+	matcher := gomega.And(assertions.PodHasContainersMatcher("sds"), assertions.PodHasContainersMatcher("istio-proxy"))
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/glooctl/istio_inject_suite.go:42]: nit: Would it make sense to define more of these contains around https://github.com/solo-io/gloo/blob/main/pkg/utils/kubeutils/names.go?

File: test/kubernetes/e2e/features/istio/gloo_gateway_no_auto_mtls_suite.go [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+	s.NoError(err, "can apply permissivePeerAuth")
+
+	// With auto mtls disabled in the mesh, the response should not contain the X-Forwarded-Client-Cert header
+	s.testInstallation.Assertions.AssertEventualCurlResponse(
+		s.ctx,
+		curlPodExecOpt,
+		[]curl.Option{
+			curl.WithHost(kubeutils.ServiceFQDN(metav1.ObjectMeta{Name: defaults.GatewayProxyName, Namespace: s.testInstallation.Metadata.InstallNamespace})),
+			curl.WithHostHeader("httpbin"),
+			curl.WithPath("/headers"),
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/istio/gloo_gateway_no_auto_mtls_suite.go:98]: nit: We string the `/` in the path, and then add it in the request to ensure consistency, so we could remove it here and it would behave the same. Not a requirement, just sharing a subtly of this method

File: test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+		err = testInstallation.UninstallIstio()
+		if err != nil {
+			t.Fatalf("failed to uninstall istio: %v", err)
+		}
+
+		testCluster.UnregisterTestInstallation(testInstallation)
+	})
+
+	// Install Istio before Gloo Gateway to make sure istiod is present before istio-proxy
+	err = testInstallation.InstallMinimalIstio(ctx)
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go:59]: How "expensive" of an operation is installing istio? In a follow up, would we ever want a top level "TestIstio" and it installs istio first, then install/unistall gloo as many times as necessary, and then uninstalls istio last?

File: test/kubernetes/testutils/assertions/pods.go [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+			g.Expect(pod).To(matcher)
+		}
+	}).
+		WithTimeout(time.Second*60).
+		WithPolling(time.Second*5).
+		Should(gomega.Succeed(), "Failed to match pod")
+}
+
+// PodHasContainersMatcher returns a GomegaMatcher that checks whether a pod has expected container with the given name.
+func PodHasContainersMatcher(containerName string) types.GomegaMatcher {
```
Comment (sam-heilbron) [inline test/kubernetes/testutils/assertions/pods.go:30]: Would it make sense to define this in here: https://github.com/solo-io/gloo/tree/main/test/gomega/matchers?

File: projects/gloo/cli/pkg/testutils/cli.go (hunk 11 above)
Comment (npolshakova) [reply projects/gloo/cli/pkg/testutils/cli.go:73]: I'll just remove it- we can use the default `istioDefaultNS`.

File: projects/gloo/cli/pkg/testutils/cli.go (hunk 12 above)
Comment (npolshakova) [reply projects/gloo/cli/pkg/testutils/cli.go:83]: Ah yep, this got copied over when switching to the new cli commands.

File: test/kubernetes/e2e/features/glooctl/istio_inject_suite.go (hunk 13 above)
Comment (npolshakova) [reply test/kubernetes/e2e/features/glooctl/istio_inject_suite.go:42]: Yep, or I think they should live near the istio commands: https://github.com/solo-io/gloo/tree/main/projects/gloo/cli/pkg/cmd/istio

File: test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go (hunk 15 above)
Comment (npolshakova) [reply test/kubernetes/e2e/gloo_gateway_edge/glooctl_istio_inject_test.go:59]: The minimal install is about as expensive as installing the Gloo helm chart, so it would be an optimization to install it first, but then all the Istio tests would need to be grouped together. I think we should consider switching this when we have to test different Istio version/different IstioOperators (ambient, revisions, etc.)

File: test/kubernetes/testutils/assertions/pods.go (hunk 16 above)
Comment (npolshakova) [reply test/kubernetes/testutils/assertions/pods.go:30]: That works!

File: projects/gloo/cli/pkg/testutils/cli.go (hunk 12 above)
Comment (sam-heilbron) [reply projects/gloo/cli/pkg/testutils/cli.go:83]: How do the tests pass with this value?

Comment (npolshakova) [reply projects/gloo/cli/pkg/testutils/cli.go]: It looks like it's not in the list of tests: `'(^TestK8sGatewayIstioAutoMtls$$|^TestK8sGatewayIstio$$|^TestClassicEdgeGateway$$)'`

Comment (npolshakova) [reply projects/gloo/cli/pkg/testutils/cli.go]: Okay added all the new suites in https://github.com/solo-io/gloo/pull/9436/commits/d75af9ae270303fe825221ae126bd99432c58718 and https://github.com/solo-io/gloo/pull/9436/commits/f665a4a33a9c475bec285150866115d954cc2b0b. I'll need to update the time once they finish running

**Review by sam-heilbron — APPROVED** [review]

**Review by jenshu — APPROVED** [review]
